### DISCUSSION TITLE
How to minimize binary library size?

### DISCUSSION THREAD

**Discussion by axxel** [zxing-cpp/zxing-cpp#715] [Q&A]
Thanks for all the work ! Trying to evaluate if I may be able to use it at work. Mostly interested in C/CPP. Since the native is CPP, that bloats up my application size, trying to keep it to a minimum.
Is there a report/ experimentation with the application sizes for C/CPP builds? 
Thanks !

_Originally posted by @pranavtriguna in https://github.com/zxing-cpp/zxing-cpp/discussions/706#discussioncomment-8375408_

I opened this new issue to not spam the "sign up" discussion. You are interested in a minimal binary size of the library?

**Comment by pranavtriguna** [comment]
Essentially, I would like to use this library to detect/decode barcode types, but refrain from using the whole library, as I may not need all the functionality. Hence the compile size is a factor, as well as integration into existing project. I read somewhere that a minimum C++ version support is needed and I'm checking on that for now. 

Not looking for a specific binary size but an estimation, if this ends up compiling in 10s or 100s of MB, so I could think to trim it down to my use. Potentially may end up cherry-pick files to compile etc., but for now going through Silky's C wrapper.

Irrespective of if I end up using this library or not, want to say thanks again for all the effort!

**Comment by pranavtriguna** [comment]
To address the topic more directly, I will look for options to partially include or exclude functionality to compile to a lower size, and potentially use C-wrapper.
Would keep porting the logic to C as a last resort, in case I can't get it to compile with C++ (my compiler may be outdated and I won't be able to udpate that since it's a part of a suite)

**Comment by axxel** [comment]
A couple things:
 * a stripped release build of the complete library is currently 1,3M
 * the c-API is now part of core library if you build it with `cmake ... -DBUILD_C_API=ON`
 * using the c-API is not making the library any smaller and you still need a c++17 compiler for your platform to use it
 * porting the code to C is ... time consuming ;)
 * you can lower the size to about half if you only need to read or write (but not both), see `BUILD_READERS`
 * there is the `CMAKE_BUILD_TYPE` which can be set to `MinSizeRel`, no idea how much that changes compared to `Release`
 * a considerable part of the binary-size is potentially the libzueci related UTF8 code which contains lots of character lookup tables. Depending on which barcode formats you want to deal with, you might not need that. But removing that will require some manual patching.

**Reply by axxel** [reply]
Regarding the future possibility to remove functionality from the code base, see #744.

**Comment by pranavtriguna** [comment]
> A couple things:
> 
> * a stripped release build of the complete library is currently 1,3M

Yeah definitely will not have that much space available :( 

> * the c-API is now part of core library if you build it with `cmake ... -DBUILD_C_API=ON`


> * using the c-API is not making the library any smaller and you still need a c++17 compiler for your platform to use it

This is what I'm trying to check. I may be able to compile the individual files, but I would need to check if the linker(s) can potentially link object files from both c/cpp in the same project. c++17 should be supported, but I'm playing with the __cplusplus to check that.

> * porting the code to C is ... time consuming ;)

Agreed, and we'll leave it at that. May take months, that's after completely understanding what the algo is trying to do. I'd much sooner learn the algo and write my own version than port hehe :) Either way, if I end up making some new code, I'll share!

> * you can lower the size to about half if you only need to read or write (but not both), see `BUILD_READERS`

This makes sense!

> * there is the `CMAKE_BUILD_TYPE` which can be set to `MinSizeRel`, no idea how much that changes compared to `Release`

Thanks! But I'm not going to build this library through the regular make, I will end up inserting relevant files into my buildsystem.

> * a considerable part of the binary-size is potentially the libzueci related UTF8 code which contains lots of character lookup tables. Depending on which barcode formats you want to deal with, you might not need that. But removing that will require some manual patching.

This helps! I'll look into it!

**Comment by axxel** [comment]
I was curious and experimented a little with the build options on clang and gcc:

|            | Release | MinRelSize |
|------------|---------|------------|
| g++-13     | 1.8M    | 1.1M       |
| clang++-16 | 1.3M    | 1.1M       |

> Yeah definitely will not have that much space available :(

How much do you have available?

**Reply by axxel** [reply]
Update:
 * I just pushed 3 changes that reduce the code size in a Release build by around 10%
 * the build size with `MinSizeRel` is 1.0MB for both clang and gcc builds where a `Release` build is 1.2M for clang and 1.4M for gcc (as a `-march=native` build on an AVX2 intel system)
 * I had some success analyzing the binary size with https://github.com/google/bloaty

**Reply by Sec-ant** [reply]
Just chime in to say this is amazing. I just updated the commit pointer in zxing-wasm:

```diff
diff --git a/README.md b/README.md
index 84a4489..29795c0 100644
--- a/README.md
+++ b/README.md
@@ -43,7 +43,7 @@ This package exports 3 subpaths: `full`, `reader` and `writer`. You can choose w
 
 ### `zxing-wasm` or `zxing-wasm/full`
 
-These 2 subpaths include functions to both read and write barcodes. The wasm binary size is ~1.17 MB.
+These 2 subpaths include functions to both read and write barcodes. The wasm binary size is ~1.06 MB.
 
 ```ts
 import {
@@ -65,7 +65,7 @@ import {
 
 ### `zxing-wasm/reader`
 
-This subpath only includes functions to read barcodes. The wasm binary size is ~930 KB.
+This subpath only includes functions to read barcodes. The wasm binary size is ~831 KB.
 
 ```ts
 import {
@@ -76,7 +76,7 @@ import {
 
 ### `zxing-wasm/writer`
 
-This subpath only includes a function to write barcodes. The wasm binary size is ~330 KB.
+This subpath only includes a function to write barcodes. The wasm binary size is ~313 KB.
 
 ```ts
 import { writeBarcodeToImageFile } from "zxing-wasm/writer";

```

**Comment by pranavtriguna** [comment]
I have around 2Mb for the final binary. Looks like I'll have to hack it away :D
Edit: I'd estimate around 500Kb at most for the barcode related stuff. I'm also going to take a guess here and say that a lot of the system files may already be present, and I'll cherry-pick files to use. I'm going to do a bit of trial and error and update if I am able to come up with a solution. 

Would not want to waste your time further on this, and will watch this repo for any contributions I could help with :)

**Reply by axxel** [reply]
Good luck!

**Comment by axxel** [comment]
The subject came up again in #1131. There are some details related to Android / libzueci / static libc++ / etc: https://github.com/zxing-cpp/zxing-cpp/issues/1131#issuecomment-5104339145
